src/parser/toplevel.lisp

```diff
@@ -554,56 +547,135 @@ consume all attributes"))))
+                                                            (mapcar #'car *package-clauses*))
+                                              :type ':help)))))
+    (funcall parser package cursor)))
+
+(defun parse-package (cursor)
+  "Parse a coalton package declaration."
+  (cursor:next-symbol cursor
+                      :require "PACKAGE"
+                      :message "package declarations must start with `package`"
+                      :missing "missing `package`")
+  (let* ((package-name
+           (cursor:next-symbol cursor
+                               :missing "missing package name"
+                               :message "package name must be a symbol"))
+         (package-doc
+           (unless (cursor:empty-p cursor)
+             (cursor:next cursor :pred #'stringp)))
+         (package
+           (make-toplevel-package :name (symbol-name package-name)
+                                  :docstring package-doc)))
+    (cursor:do-every cursor (alexandria:curry 'parse-package-clause package))
+    package))
 
-    (unless package
-      (setf package (make-package package-name :use '("COALTON" "COALTON-PRELUDE"))))
+;; Empty package for reading (package) forms
 
-    package))
+(defpackage #:coalton-impl/parser/%defpackage
+  (:documentation "The sole purpose of this package is to provide storage for symbols that are interned while Coalton package forms are read.")
+  (:use))
 
+(defmacro with-parser-package (&body body)
+  `(let ((*package* (find-package 'coalton-impl/parser/%defpackage)))
+     (unwind-protect
+          (progn ,@body)
+       (do-symbols (symbol)
+         (unintern symbol)))))
+
+(defun read-toplevel-package (stream file)
+  "Read and parse a Coalton toplevel package form."
+  (with-parser-package
+    (handler-bind ((cursor:syntax-error
+                     (lambda (syntax-error)
+                       (cursor:parse-error file
+                                           "Malformed package declaration"
+                                           syntax-error))))
+      (multiple-value-bind (form presentp)
+          (maybe-read-form stream *coalton-eclector-client*)
+        (unless presentp
+          (cursor:span-error (cons (- (file-position stream) 2)
+                                   (- (file-position stream) 1))
+                      "missing package form"))
+        (parse-package (cursor:make-cursor form))))))
+
+(defun make-defpackage (package)
+  "Generate a Lisp defpackage form from a Caolton toplevel-package structure."
+  `(defpackage ,(toplevel-package-name package)
+     (:use "COALTON" ,@(toplevel-package-import package))
+     ,@(mapcar (lambda (form)
+                 `(:import-from ,@form))
+               (toplevel-package-import-from package))
+     (:local-nicknames ,@(toplevel-package-import-as package))
+     (:export ,@(toplevel-package-export package))))
+
+(defun lisp-package (package)
```

Comment:
Can we use `make-package` instead of `eval`?

Reply:
Only reason not to is `make-package` doesn't support package-local nicknames. There may be a good opportunity to revisit package invariants when specifying how redefinitions work during interactive development. I'm thinking in particular of how clunky changes to package exports are, when working with common lisp.